### PR TITLE
Adding Charm plus plus compiler for HPC workloads

### PR DESCRIPTION
**Pull request by deep1712**

### CODE REVIEW

**Comment by yangpanMS** [comment]
Could you help me understand why we need charm++ compiler and which workload is this going to support?

File: src/VirtualClient/VirtualClient.Dependencies/MsMPIInstallation.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            this.packageManager = dependencies.GetService<IPackageManager>();
+            this.processManager = this.systemManagement.ProcessManager;
+        }
+
+        /// <summary>
+        /// Installs the MS-MPI.
+        /// </summary>
+        protected async override Task ExecuteAsync(EventContext telemetryContext, CancellationToken cancellationToken)
+        {
+            if (this.Platform == PlatformID.Win32NT)
```
Comment (brdeyo) [inline src/VirtualClient/VirtualClient.Dependencies/MsMPIInstallation.cs:43]: Implement the IsSupported method. If we only support Windows, the installation should no-op

File: src/VirtualClient/VirtualClient.Dependencies/CompilerInstallation.cs [hunk 2]
```diff
@@ -254,6 +265,25 @@ private async Task InstallAoccAsync(string aoccVersion, EventContext telemetryCo
...
+            string lockFile = this.PlatformSpecifics.Combine(Environment.CurrentDirectory, "charmcompilersuccess.lock");
+            if (!fileInterface.Exists(lockFile))
+            {
+                charmplusplusVersion = (string.IsNullOrEmpty(charmplusplusVersion)) ? "latest" : charmplusplusVersion;
+                await this.ExecuteCommandAsync("wget", $"https://charm.cs.illinois.edu/distrib/charm-{charmplusplusVersion}.tar.gz -O charm.tar.gz", Environment.CurrentDirectory, telemetryContext, cancellationToken)
+                    .ConfigureAwait(false);
+                await this.ExecuteCommandAsync("tar", $"-xzf charm.tar.gz", Environment.CurrentDirectory, telemetryContext, cancellationToken)
+                    .ConfigureAwait(false);
+                string charmPath = Directory.GetDirectories(Environment.CurrentDirectory, "charm-v*").FirstOrDefault();
+                await this.ExecuteCommandAsync("./build", "charm++ netlrts-linux-x86_64 --with-production -j4", workingDirectory: charmPath, telemetryContext, cancellationToken)
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Dependencies/CompilerInstallation.cs:281]: I'm not seeing a check for x64 vs ARM64 here

**Comment by yangpanMS** [comment]
Could you add documents to https://microsoft.github.io/VirtualClient/docs/dependencies/0032-compiler-installation/ ?

File: src/VirtualClient/VirtualClient.Dependencies/MsMPIInstallation.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    using Microsoft.Extensions.DependencyInjection;
+    using VirtualClient.Common;
+    using VirtualClient.Common.Extensions;
+    using VirtualClient.Common.Telemetry;
+    using VirtualClient.Contracts;
+
+    /// <summary>
+    /// Installation component for the MS-MPI.
+    /// </summary>
+    public class MSMPIInstallation : VirtualClientComponent
```
Comment (brdeyo) [inline src/VirtualClient/VirtualClient.Dependencies/MsMPIInstallation.cs:17]: Name of file and name of class should be identical. Use MsmpiInstallation as the name following standard .NET naming conventions.

File: src/VirtualClient/VirtualClient.Packaging/win-x64/visualstudiocruntime/msvcr100.dll
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Packaging/win-x64/visualstudiocruntime/msvcr100.dll:1]: Could you either put a doc next to it with where you got it, or include that in vcpkg

File: src/VirtualClient/VirtualClient.Dependencies/MsMPIInstallation.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                        telemetryContext,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+
+                    fileInterface.Create(lockFile);
+                }
+            }
+        }
+
+        private async Task InstallComponentAsync(string msiPath, EventContext telemetryContext, CancellationToken cancellationToken)
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Dependencies/MsMPIInstallation.cs:79]: We have an extension method for running, logging, throwing altogether in VC component now.

File: src/VirtualClient/VirtualClient.Dependencies/MsMPIInstallation.cs (hunk 3 above)
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Dependencies/MsMPIInstallation.cs:17]: Please also add a document for this dependency class

File: src/VirtualClient/VirtualClient.Dependencies/MsmpiInstallation.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            DependencyPath workloadPackage = await this.packageManager.GetPlatformSpecificPackageAsync(
+                this.PackageName, this.Platform, this.CpuArchitecture, CancellationToken.None)
+                .ConfigureAwait(false);
+
+            string packageDirectory = workloadPackage.Path;
+
+            string lockFile = this.PlatformSpecifics.Combine(packageDirectory, MsmpiInstallation.LockFileName);
+            IFile fileInterface = this.systemManagement.FileSystem.File;
+
+            if (!fileInterface.Exists(lockFile))
```
Comment (brdeyo) [inline src/VirtualClient/VirtualClient.Dependencies/MsmpiInstallation.cs:53]: Use state to track changes on the system that need to be remembered on a second go. Deviation from our process prevents us from being able to do a reset in a clean manner.

File: website/docs/dependencies/0053-install-msmpi.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+## Example
+The following section describes the parameters used by the individual component in the profile.
+
+```json
+    {
+      "Type": "DependencyPackageInstallation",
+      "Parameters": {
+        "Scenario": "DownloadMSMPIPackage",
+        "BlobContainer": "packages",
+        "BlobName": "msmpi10.1.2.zip",
```
Comment (brdeyo) [inline website/docs/dependencies/0053-install-msmpi.md:26]: Rename the package: msmpi.10.1.2.zip

The version of the toolset should be separate from the name

File: src/VirtualClient/VirtualClient.Dependencies/MsmpiInstallation.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                string msmpiBin = this.PlatformSpecifics.Combine(packageDirectory, "bin");
+                this.SetEnvironmentVariable(EnvironmentVariable.PATH, msmpiBin, EnvironmentVariableTarget.Machine, true);
+
+                FirewallEntry firewallEntry = new FirewallEntry(
+                $"Virtual Client: Allow {this.PlatformSpecifics.Combine(msmpiBin, "smpd.exe")}",
+                $"Allows smpd to Communicate in Multiple Machine Scenario.",
+                this.PlatformSpecifics.Combine(msmpiBin, "smpd.exe"));
+
+                await this.systemManagement.FirewallManager.EnableInboundAppAsync(firewallEntry, cancellationToken)
+                    .ConfigureAwait(false);
```
Comment (brdeyo) [inline src/VirtualClient/VirtualClient.Dependencies/MsmpiInstallation.cs:64]: Let's remove the .ConfigureAwait(false); from all of the logic. Yang and I have confirmed it is not necessary and does nothing when used in non-U/X logic. The .NET team has removed the requirement for it. This means that .ConfigureAwait(true) and .ConfigureAwait(false) both do exactly the same thing in VC code. We can thus remove the construct and cleanup our code.

File: website/docs/dependencies/0053-install-msmpi.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+| PackageName   | Yes          | BlobStorage PackageName.                                |
+
+## Example
+The following section describes the parameters used by the individual component in the profile.
+
+```json
+    {
+      "Type": "DependencyPackageInstallation",
+      "Parameters": {
+        "Scenario": "DownloadMSMPIPackage",
```
[the remark this answers is not in the document]
Comment (deep1712) [reply website/docs/dependencies/0053-install-msmpi.md:24]: I think we have mixed the meaning in VC.

Download: It means to copy/drop the binaries/packages to a system from Internet.
Install: It means to do all the setup in the system and make it ready to run.

That means nuget package is downloaded and extracted.

Or am I miss interpreting the terminology?

[the remark this answers is not in the document]
Comment (brdeyo) [reply website/docs/dependencies/0053-install-msmpi.md:24]: The term "Install" has a wide range of meanings. Your definition above is but 1 of them. This is more about consistency with other profiles. We have been using the terminology "InstallXYZPackage". We can decide in the future to change the name of the scenario, but for now stick with the convention.

**Review by brdeyo — APPROVED** [review]

**Pull request merged**